source/numerics.tex

```diff
@@ -3061,6 +3067,239 @@
+  public:
+    // types
+    using result_type = UIntType;
+
+    // engine characteristics
+    static constexpr size_t word_size = w;
+    static constexpr size_t word_count = n;
+    static constexpr size_t round_count = r;
+    static constexpr array<result_type, @\exposid{array-size}@> multipliers;
+    static constexpr array<result_type, @\exposid{array-size>}@ round_consts;
+    static constexpr result_type min() { return 0; }
+    static constexpr result_type max() { return m - 1; }
+    static constexpr result_type default_seed = 20111115u;
+
+    // constructors and seeding functions
+    philox_engine() : philox_engine(default_seed) {}
+    explicit philox_engine(result_type value);
+    template<class Sseq> explicit philox_engine(Sseq& q);
+    void seed(result_type value = default_seed);
+    template<class Sseq> void seed(Sseq& q);
+
+    void set_counter(const array<result_type, n>& counter);
+
+    // equality operators
+    friend bool operator==(const philox_engine& x, const philox_engine& y);
+
+    // generating functions
+    result_type operator()();
+    void discard(unsigned long long z);
+
+    // inserters and extractors
+    template<class charT, class traits>
+      friend basic_ostream<charT, traits>&
+        operator<<(basic_ostream<charT, traits>& os, const philox_engine& x);
+    template<class charT, class traits>
+      friend basic_istream<charT, traits>&
+        operator>>(basic_istream<charT, traits>& is, philox_engine& x);
+  };
+}
+\end{codeblock}
+
+\pnum
+\mandates
+\begin{itemize}
+\item \tcode{sizeof...(consts) == n} is \tcode{true}, and
+\item \tcode{n == 2 || n == 4 || n == 8 || n == 16} is \tcode{true}, and
+\item \tcode{0 < r} is \tcode{true}, and
+\item \tcode{0 < w \&\& w <= numeric_limits<UIntType>::digits} is \tcode{true}.
+\end{itemize}
+
+\pnum
+The template parameter pack \tcode{consts} represents
+the $M_k$ and $C_k$ constants which are grouped as follows:
+$[ M_0, C_0, M_1, C_1, M_2, C_2, \dotsc, M_{n/2 - 1}, C_{n/2 - 1} ]$.
+
+\pnum
+The textual representation consists of the values of
+$K_0, \dotsc, K_{n/2 - 1}, X_{0}, \dotsc, X_{n - 1}, i$, in that order.
+\begin{note}
+The stream extraction operator can reconstruct $Y$ from $K$ and $X$, as needed.
+\end{note}
+
+\indexlibraryctor{philox_engine}
+\begin{itemdecl}
+explicit philox_engine(result_type value);
+\end{itemdecl}
+
+\begin{itemdescr}
+\pnum
+\effects
+Sets the $K_0$ element of sequence $K$ to $\tcode{value} \mod 2^w$.
+All elements of sequences $X$ and $K$ (except $K_0$) are set to \tcode{0}.
+The value of $i$ is set to $n - 1$.
+\end{itemdescr}
+
+\indexlibraryctor{philox_engine}
+\begin{itemdecl}
+template<class Sseq> explicit philox_engine(Sseq& q);
+\end{itemdecl}
+
+\begin{itemdescr}
+\pnum
+\effects
+With $p = \left\lceil w / 32 \right\rceil$ and
+an array (or equivalent) \tcode{a} of length $(n/2) \cdot p$,
+invokes \tcode{q.generate(a + 0, a + n / 2 * $p$)} and
```

Comment:
"array (or equivalent)" is a bit vague.

```suggestion
a random access iterator \tcode{a} such that \tcode{a} + \range{0}{$(n/2) \cdot p$} is valid,
invokes \tcode{q.generate(a + 0, a + n / 2 * $p$)} and
```

But maybe this should be an LWG issue, or maybe it's fine. I can't remember if this was discussed during the review.

Reply:
That is an exact wording used in all the other engines. If changed, better be changed consistently everywhere.

Reply:
Ah ok! We should leave it then